src/java_bytecode/java_types.cpp

```diff
@@ -167,6 +167,243 @@ exprt java_bytecode_promotion(const exprt &expr)
     return typecast_exprt(expr, new_type);
 }
 
+std::vector<typet> parse_list_types(
+  const std::string src,
+  const std::string class_name_prefix,
+  const char opening_bracket,
+  const char closing_bracket);
+
+/// Take a list of generic arguments and parse them into the generic type.
+/// \param generic_type [out]: The existing generic type to add the information
+///   to
+/// \param parameters: The string representing the generic arguments for a
+///   signature. For example <TT;Ljava/lang/Foo;> (including wrapping angle
+///   brackets).
+/// \param class_name_prefix: The name of the class to use to prefix any found
+///   generic types
+void add_generic_type_information(
+  java_generic_typet &generic_type,
+  const std::string &parameters,
+  const std::string &class_name_prefix)
+{
+  PRECONDITION(parameters[0] == '<');
+  PRECONDITION(parameters[parameters.size() - 1] == '>');
+
+  // parse contained types, can be either type variables, starting with T
+  // or instantiated types
+  const std::vector<typet> &params =
+    parse_list_types(parameters, class_name_prefix, '<', '>');
+
+  CHECK_RETURN(params.size() > 0); // We should have at least one generic param
+
+  // take these types - they should either be java_generic_parameters, in which
+  // case they can be directly added to the generic_type
+  // otherwise they should be wrapped in a java_generic_inst_parametert
+
+  std::transform(
+    params.begin(),
+    params.end(),
+    std::back_inserter(generic_type.generic_type_variables()),
+    [](const typet &type) -> java_generic_parametert {
+      if(is_java_generic_parameter(type))
+      {
+        const java_generic_parametert &gen_type =
+          to_java_generic_parameter(type);
+        return gen_type;
+      }
+      else
+      {
+        INVARIANT(
+          is_reference(type), "All generic parameters should be references");
+        java_generic_inst_parametert inst_type(to_symbol_type(type.subtype()));
+        return inst_type;
+      }
+    });
+}
+
+/// Take a signature string and remove everything in angle brackets allowing
+/// for the type to be parsed normally.
+/// \param src: The input string
+/// \return The input string with everything between angle brackets removed
+std::string erase_type_arguments(const std::string &src)
+{
+  std::string class_name = src;
+  std::size_t f_pos = class_name.find('<', 1);
+
+  while(f_pos != std::string::npos)
+  {
+    std::size_t e_pos = find_closing_delimiter(class_name, f_pos, '<', '>');
+    if(e_pos == std::string::npos)
+      throw unsupported_java_class_signature_exceptiont("recursive generic");
```

Comment:
I don't understand this, wouldn't this return `npos` if no corresponding closing `>` can be found, not if it is recursive generic?

Reply:
You are right - however this case happens precisely when we have a recursive generic type. Given ```java_generic_type_from_string(<T:LSimpleRecursiveGeneric<TT;>;>Ljava/lang/Object;)``` calls erase type arguments with `T:LSimpleRecursiveGeneric<TT`